### ISSUE TITLE
DHCP not working due to bad udp checksum - TX offloading problem

### ISSUE THREAD

**Issue by localpref** [Metaswitch/calico#40]
My CirrOS based virtual machines  `Linux cirros 3.2.0-68-virtual #102-Ubuntu SMP Tue Aug 12 22:14:39 UTC 2014 x86_64 GNU/Linux` ignore the DHCP reply sent by dnsmasq. When tcpdumping on the VM's tap interface, I can see the DHCP request AND the DHCP reply, but the reply is not accepted.

It appears to be down to a problem with checksum generation, which usually only happens on the physical interface (offloading), but with Calico all DHCP happens inside the same compute node and only goes from the dnsmasq ns to the relevant tap ports...

A workaround that helped me make my instances accept the DHCP reply is adding the following iptables/mangle entry on my compute node:

`iptables -A POSTROUTING -t mangle -p udp --dport bootpc -d 100.64.0.0/10 -j CHECKSUM --checksum-fill`

the one odd thing that I see in the attached tcpdump files is that the IP source address is that of the compute node (10.13.0.140), while the DHCP server address contained is that of the virtual gateway 100.64.0.1, and maybe that's where the wrong checksum comes from?

however, I think the problem needs to be debugged further - maybe there is a way to signal that there is no offloading support, and maybe dnsmasq needs to be patched to properly calculate checksums on request?
Or is there something more sinister going on here?

It appears as though I am hitting something related to
http://article.gmane.org/gmane.linux.kernel/1003853

more info for a busybox patch from 2012:
http://lists.busybox.net/pipermail/busybox/2012-September/078348.html

more info:
https://github.com/mseknibilel/OpenStack-Folsom-Install-guide/issues/14
https://bugzilla.novell.com/show_bug.cgi?id=668194

tcpdump with broken reply:
12:48:43.085347 IP (tos 0xc0, ttl 64, id 61166, offset 0, flags [none], proto UDP (17), length 336)
    10.13.0.140.67 > 100.64.0.4.68: [bad udp cksum 0x702a -> 0x342f!] BOOTP/DHCP, Reply, length 308, xid 0xc8b8dc3d, Flags [none] (0x0000)
	  Your-IP 100.64.0.4
	  Server-IP 100.64.0.1
	  Client-Ethernet-Address fa:16:3e:57:03:66
	  Vendor-rfc1048 Extensions
	    Magic Cookie 0x63825363
	    DHCP-Message Option 53, length 1: Offer
	    Server-ID Option 54, length 4: 100.64.0.1
	    Lease-Time Option 51, length 4: 86400
	    RN Option 58, length 4: 43200
	    RB Option 59, length 4: 75600
	    Subnet-Mask Option 1, length 4: 255.192.0.0
	    BR Option 28, length 4: 100.127.255.255
	    Domain-Name-Server Option 6, length 4: 100.64.0.1
	    Domain-Name Option 15, length 14: "openstacklocal"
	    Default-Gateway Option 3, length 4: 100.64.0.1

**Comment by neiljerram** [comment]
Similar report from Spike Curtis:

I ran into this bug: https://bugs.launchpad.net/ubuntu/+source/isc-dhcp/+bug/930962 , which prevented DHCP offers from being accepted by guest VMs.   While this isn’t a Calico bug, it’s worth pointing out in the troubleshooting section, since I didn’t run afoul of it with Neutron + OVS.  It seems to affect OpenStack deployments where the DHCP server is on the same host machine as the guest VM---which how Calico is configured to work!
[...]
The checksum is wrong on the DHCP Server to VM packets, e.g. the DHCP Offer.
[...]
I’m pretty sure the problem was with virtio, as described in the bug I linked to, since it was resolved when I set the _virtio_ option in nova.conf to false.  I didn’t make any changes to iptables.

**Comment by fasaxc** [comment]
I don't remember seeing that rule on more recent versions of Felix, we should check it's still there and then close this out (or update the docs).

**Comment by nbartos** [comment]
FYI I found out this is only necessary when using the vhost_net kernel module on the hypervisor when using older guests that use virtio.

I'm currently installing the rule (after enabling vhost_net) in the OUTPUT mangle chain, since it only needs to be done for packets coming from the hypervisor to the VM.  I believe putting it in the POSTROUTING mangle chain will result in more packets being targeted, e.g. packets that are routed in.  Since it's not a problem for routed packets, OUTPUT should be fine (and better to have a more targeted rule).

https://bugzilla.redhat.com/show_bug.cgi?id=910619#c6

**Comment by fasaxc** [comment]
Thanks for the heads up @nbartos.

**Comment by neiljerram** [comment]
FYI there's a current thread on this on openstack-dev: http://lists.openstack.org/pipermail/openstack-dev/2015-June/065394.html.  Apparently the iptables fix was merged into the Neutron DHCP agent, before the Kilo release.

**Comment by DaveLangridge** [comment]
I just hit this on RHEL7 calico + kilo. Compute node was working OK, then I rebooted. After reboot the checksum rule was missing, and cirros VMs didn't get their IP.

**Comment by neiljerram** [comment]
I suspect this might be a genuine bug connected with the 'use_namespaces' option.  Please could you try, in `/etc/neutron/dhcp_agent.ini` on a compute node, changing

    # use_namespaces = True

to 

    use_namespaces = False

Then reboot and see if that solves the problem for that compute node.

The situation here is that for Calico operation we always want the DHCP agent to behave as though use_namespaces is False, and we've implemented that by adding an extra condition (self.bridged) to various places that test for self.conf.use_namespaces.  However it looks like the Kilo release has added several new places that test for self.conf.use_namespaces - specifically including the place that sets up the iptables rules for checksumming - and our patch does not cover those new places.

**Comment by DaveLangridge** [comment]
Didn't help. I set this to 'False' and rebooted, the started the VM. No IP.

**Comment by neiljerram** [comment]
Ah, there's a traceback in dhcp-agent.log:

    2015-06-08 16:29:01.208 2099 ERROR neutron.agent.dhcp.agent [-] Unable to enable dhcp for 7513f4e6-1b25-406c-b22c-96d5a48078aa.
    2015-06-08 16:29:01.208 2099 TRACE neutron.agent.dhcp.agent Traceback (most recent call last):
    2015-06-08 16:29:01.208 2099 TRACE neutron.agent.dhcp.agent   File "/usr/lib/python2.7/site-packages/neutron/agent/dhcp/agent.py", line 128, in call_driver
    2015-06-08 16:29:01.208 2099 TRACE neutron.agent.dhcp.agent     getattr(driver, action)(**action_kwargs)
    2015-06-08 16:29:01.208 2099 TRACE neutron.agent.dhcp.agent   File "/usr/lib/python2.7/site-packages/neutron/agent/linux/dhcp.py", line 201, in enable
    2015-06-08 16:29:01.208 2099 TRACE neutron.agent.dhcp.agent     interface_name = self.device_manager.setup(self.network)
    2015-06-08 16:29:01.208 2099 TRACE neutron.agent.dhcp.agent   File "/usr/lib/python2.7/site-packages/neutron/agent/linux/dhcp.py", line 994, in setup
    2015-06-08 16:29:01.208 2099 TRACE neutron.agent.dhcp.agent     namespace=network.namespace)
    2015-06-08 16:29:01.208 2099 TRACE neutron.agent.dhcp.agent   File "/usr/lib/python2.7/site-packages/neutron/agent/linux/interface.py", line 460, in plug
    2015-06-08 16:29:01.208 2099 TRACE neutron.agent.dhcp.agent     ns_veth = ip.add_dummy(device_name, namespace2=namespace)
    2015-06-08 16:29:01.208 2099 TRACE neutron.agent.dhcp.agent   File "/usr/lib/python2.7/site-packages/neutron/agent/linux/ip_lib.py", line 150, in add_dummy
    2015-06-08 16:29:01.208 2099 TRACE neutron.agent.dhcp.agent     return IPDevice(name1, self.root_helper, self.namespace)
    2015-06-08 16:29:01.208 2099 TRACE neutron.agent.dhcp.agent AttributeError: 'IPWrapper' object has no attribute 'root_helper'

**Comment by neiljerram** [comment]
Addressed by https://github.com/Metaswitch/calico-neutron/pull/12

**Comment by neiljerram** [comment]
Possibly what we should now do - to fix this properly for all platforms, and close this issue - is (in the core Calico repository):

- backport the upstream Kilo fix to our Icehouse and Juno versions of Neutron

- remove the CHECKSUM and iptables-save lines from debian/calico-compute.postinst and rpm/calico.spec.

Thoughts?

**Comment by nbartos** [comment]
I agree in general, but I would definitely use a slightly different rule than the one in openstack.  The one currently in openstack will mangle packets routed in from other sources (VMs or outside), not just ones going from the host to the VM.  Instead of using the PREROUTING chain in the mangle table, the OUTPUT chain should be used.  Additionally, an interface should be specified to limit the scope.  For example:

iptables -t mangle -A OUTPUT -o tap+ -p udp -m udp --dport 68 -j CHECKSUM --checksum-fill

This is actually the rule I have been using in our Calico deployment with success.

It would be good if the openstack rule could be disabled in some way so unnecessary packets don't get mangled.

**Comment by neiljerram** [comment]
Thanks @nbartos.  I think the right thing would be to get those changes upstream in OpenStack, if we can motivate them sufficiently.

(Note: it's POSTROUTING that it currently used by OpenStack and our install time stuff, not PREROUTING.)

For OUTPUT instead of POSTROUTING: Yes, this sounds to me like an unambiguous win, as there can be no need for forwarded packets to hit this rule.  I've commented at https://review.openstack.org/#/c/148718/ - let's see if anyone replies.

I'm less sure about filtering on interface name, because that seems like a relatively expensive check, compared to --dport 68, and because it really isn't possible for DHCP packets, in a Calico setup, to go anywhere other than the TAP interfaces; this is already controlled by the fact that we invoke dnsmasq with --bridge-interface=ns-XXX,tap*.  Is there anything else that I'm missing here?

**Comment by nbartos** [comment]
Thanks for catching my typo on the chain name.  Yes I agree changing the chain is the most important thing, the interface less so.  I'm just in the habit of making things of this nature as specific as possible, to avoid unintentional consequences.  While I can't at the moment think of a case where things would break without specifying the interface, I've definitely been bitten by issues in the past when the rules were not as specific as they should have been.

**Comment by neiljerram** [comment]
Related bug: https://bugs.launchpad.net/neutron/+bug/1465280

**Comment by fasaxc** [comment]
@neiljerram I think we discussed that this has been fixed in kilo so our next action is to port that fix to Juno/Icehouse and remove our own version.

**Comment by neiljerram** [comment]
@fasaxc Agreed

**Comment by matthewdupre** [comment]
I've split #709 out from this to cover narrowing the rule down.  I'll backport the Kilo fix now and then close this one down.

**Comment by matthewdupre** [comment]
Kilo neutron fix backported, will be released in 0.27.  Closing this one, although #709 will remain open.

**Comment by elico** [comment]
@wido on my KVM host I added this rule to a dhcp server machine:

> iptables -A POSTROUTING -t mangle -p udp --dport 68 -j CHECKSUM --checksum-fill

 since a debian 8 jessie guest dhcp client on the same hypervisor couldn't verify the checksum... sasince there isn't.

**Comment by fasaxc** [comment]
@elico this is a long-closed issue, if you think there's still a bug in this area, please can you open a new issue?

**Comment by elico** [comment]
@fasaxc I do not know if this issue exist or not and I believe it got fixed but I added a note that it still exists and with a specific OS version that I have tested on a regular KVM host...
I will try to test with debian 9 and other OS versions and if I will find the issue still exists I will open a new bug report.

**Comment by analistamarcioprado** [comment]
Good Morning. We are facing the same problem. But we imagine that the problem is not the DHCP itself.

In the Neutral documentation it says: "Computing hosts should always have the appropriate NICs set to promiscuous mode."

This is necessary for the host to receive frames addressed to other MACs (from a VM hosting on it for example). This is required for VMs to be able to respond to an IP offer.

Perhaps this may be more effectively related to DHCP malfunction.

I came to manually configure the IPs on the VMs, and yet the (intermittent) communication problem continued ... Only after several reboots of all cloud hosts did the VMs re-communicate.

**Comment by neiljerram** [comment]
@analistamarcioprado How are you installing Felix?  Have you reviewed the comments above?
